runners/spark/src/main/java/org/apache/beam/runners/spark/SparkRunner.java

```diff
@@ -242,8 +227,54 @@ public void visitPrimitiveTransform(TransformTreeNode node) {
       doVisitTransform(node);
     }
 
-    protected abstract <TransformT extends PTransform<? super PInput, POutput>> void
-        doVisitTransform(TransformTreeNode node);
+    <TransformT extends PTransform<? super PInput, POutput>>
+    void doVisitTransform(TransformTreeNode node) {
+      @SuppressWarnings("unchecked")
+      TransformT transform = (TransformT) node.getTransform();
+      @SuppressWarnings("unchecked")
+      Class<TransformT> transformClass = (Class<TransformT>) (Class<?>) transform.getClass();
+      PCollection.IsBounded isBounded = isNodeBounded(node);
+      LOG.info("Translating {} as {}", transform, isBounded);
+      @SuppressWarnings("unchecked") TransformEvaluator<TransformT> evaluator =
+          translator.translate(transformClass, isBounded);
+      LOG.info("Evaluating {}", transform);
+      AppliedPTransform<PInput, POutput, TransformT> appliedTransform =
+          AppliedPTransform.of(node.getFullName(), node.getInput(), node.getOutput(), transform);
+      ctxt.setCurrentTransform(appliedTransform);
+      evaluator.evaluate(transform, ctxt);
+      ctxt.setCurrentTransform(null);
+    }
+
+    /** Determine if this Node belongs to a Bounded branch of the pipeline, or Unbounded. */
+    private PCollection.IsBounded isNodeBounded(TransformTreeNode node) {
+      // usually, the input determines PCollection to apply the next transformation to is BOUNDED
+      // or UNBOUNDED, meaning RDD/DStream - thus mapping to Streaming/EvaluationContext.
+      Collection<? extends PValue> pValues;
+      PInput pInput = node.getInput();
+      if (pInput instanceof PBegin) {
+        // in case of a PBegin, it's the output.
```

Comment:
OK, now I see it a bit more clearly. Since the proposed portable model is such that only a PCollection is a value, there is now some drift between PValue and the model-level concept. I had thought that a PCollectionView would expand() to the PCollection that it is viewing, but it does not. So this LGTM.